### ISSUE TITLE
compilation errors when cross-building for 32bit systems

### ISSUE THREAD

**Issue by chrysos349** [f-koehler/KTailctl#46]
Hi,

Here is the error when trying to cross-build your app for `i686`. The same error when compiling for `armv7l`.
```
error: right operand of shift expression '(1 << 40)' is greater than or equal to the precision 32 of the left operand [-fpermissive]
```

Can your app be cross-compiled or compiled natively for 32bit archs?

P.S. Here is the excerpt from the log
```
[27/28] Building CXX object src/CMakeFiles/ktailctl.dir/util.cpp.o
FAILED: src/CMakeFiles/ktailctl.dir/util.cpp.o
/usr/lib/ccache/bin/g++ -DKCOREADDONS_LIB -DKGUIADDONS_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_QMLMODELS_LIB -DQT_QML_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICK_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_XML_LIB -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=600 -I/builddir/ktailctl-0.8.0/build/src -I/builddir/ktailctl-0.8.0/src -I/builddir/ktailctl-0.8.0/build/src/ktailctl_autogen/include -I/builddir/ktailctl-0.8.0/build/src/src -I/builddir/ktailctl-0.8.0/build/lib -I/builddir/ktailctl-0.8.0/build/tailwrap -I/builddir/ktailctl-0.8.0/tailwrap -I /usr/include/KF5/KConfig -I /usr/include/KF5/KConfigCore -I /usr/include/KF5 -I /usr/include/qt5 -I /usr/include/qt5/QtCore -I /usr/lib/qt5/mkspecs/linux-g++ -I /usr/include/KF5/KConfigGui -I /usr/include/qt5/QtGui -I /usr/include/qt5/QtXml -I /usr/include/KF5/KCoreAddons -I /usr/include/KF5/KGuiAddons -I /usr/include/KF5/KI18n -I /usr/include/qt5/QtQml -I /usr/include/qt5/QtNetwork -I /usr/include/qt5/QtQuick -I /usr/include/qt5/QtQmlModels -I /usr/include/qt5/QtQuickControls2 -I /usr/include/qt5/QtSvg -I /usr/include/qt5/QtWidgets -DNDEBUG -fno-PIE -fno-PIE -mtune=i686 -O2    -ffile-prefix-map=/builddir/ktailctl-0.8.0/build=. -fno-operator-names -fno-exceptions -Wall -Wextra -Wcast-align -Wchar-subscripts -Wformat-security -Wno-long-long -Wpointer-arith -Wundef -Wnon-virtual-dtor -Woverloaded-virtual -Werror=return-type -Werror=init-self -Wvla -Wdate-time -Wsuggest-override -Wlogical-op -fdiagnostics-color=always -std=c++17 -flto=auto -fno-fat-lto-objects -fvisibility=hidden -fvisibility-inlines-hidden -fPIC -MD -MT src/CMakeFiles/ktailctl.dir/util.cpp.o -MF src/CMakeFiles/ktailctl.dir/util.cpp.o.d -o src/CMakeFiles/ktailctl.dir/util.cpp.o -c /builddir/ktailctl-0.8.0/src/util.cpp
../src/util.cpp: In function 'QString formatCapacityHumanReadable(long int)':
../src/util.cpp:42:42: warning: left shift count >= width of type [-Wshift-count-overflow]
   42 |     static constexpr long limits[] = {1L << 40, 1L << 30, 1L << 20, 1L << 10, 1L};
      |                                       ~~~^~~~~
../src/util.cpp:42:42: error: right operand of shift expression '(1 << 40)' is greater than or equal to the precision 32 of the left operand [-fpermissive]
../src/util.cpp: In function 'QString formatSpeedHumanReadable(double)':
../src/util.cpp:54:64: warning: left shift count >= width of type [-Wshift-count-overflow]
   54 |     static constexpr double limits[] = {static_cast<double>(1L << 40),
      |                                                             ~~~^~~~~
../src/util.cpp:54:64: error: right operand of shift expression '(1 << 40)' is greater than or equal to the precision 32 of the left operand [-fpermissive]
../src/util.cpp: In function 'QString formatDurationHumanReadable(const QDateTime&, const QDateTime&)':
../src/util.cpp:70:39: warning: integer overflow in expression of type 'long int' results in '1187194880' [-Woverflow]
   70 |         {365l * 30l * 24l * 60l * 60l * 1000l, 34 * 30l * 60l * 60l * 1000l, 24l * 60l * 60l * 1000l, 60l * 60l * 1000l, 60l * 1000l, 1000l};
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
../src/util.cpp:70:69: warning: integer overflow in expression of type 'long int' results in '-622967296' [-Woverflow]
   70 |         {365l * 30l * 24l * 60l * 60l * 1000l, 34 * 30l * 60l * 60l * 1000l, 24l * 60l * 60l * 1000l, 60l * 60l * 1000l, 60l * 1000l, 1000l};
      |                                                ~~~~~~~~~~~~~~~~~~~~~^~~~~~~
ninja: build stopped: subcommand failed.
```

**Comment by f-koehler** [comment]
I think I could modify the code such that it compiles on 32bit. The problem seems to be very limited to some literals in utility functions used to format units.

**Comment by f-koehler** [comment]
@chrysos349 Can you try the `f-koehler/issue46` branch? I fixed the integer literals to use the `ll` suffix which should yield 64bit integers even when compiling for 32bit. I do not have a good environment to test this atm.

**Comment by chrysos349** [comment]
@f-koehler Tried to cross-build for i686, using the branch `f-koehler/issue46`. I got this error:

```
error: narrowing conversion of '1099511627776' from 'long long int' to 'long int' [-Wnarrowing]
```

So I added `-Wno-narrowing` to compilation flags, and it worked! I succesfully compiled your app for i686, armv7l, and armv6l-musl. Thanks for your fix!

P.P.S. Here is the excerpt from the log for your information:

```
FAILED: src/CMakeFiles/ktailctl.dir/util.cpp.o
/usr/lib/ccache/bin/g++ -DKCOREADDONS_LIB -DKGUIADDONS_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_QMLMODELS_LIB -DQT_QML_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICK_LIB -DQT_SVG_LIB -DQT_WIDGETS_LIB -DQT_XML_LIB -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=600 -I/builddir/ktailctl-0.8.0/build/src -I/builddir/ktailctl-0.8.0/src -I/builddir/ktailctl-0.8.0/build/src/ktailctl_autogen/include -I/builddir/ktailctl-0.8.0/build/src/src -I/builddir/ktailctl-0.8.0/build/lib -I/builddir/ktailctl-0.8.0/build/tailwrap -I/builddir/ktailctl-0.8.0/tailwrap -I /usr/include/KF5/KConfig -I /usr/include/KF5/KConfigCore -I /usr/include/KF5 -I /usr/include/qt5 -I /usr/include/qt5/QtCore -I /usr/lib/qt5/mkspecs/linux-g++ -I /usr/include/KF5/KConfigGui -I /usr/include/qt5/QtGui -I /usr/include/qt5/QtXml -I /usr/include/KF5/KCoreAddons -I /usr/include/KF5/KGuiAddons -I /usr/include/KF5/KI18n -I /usr/include/qt5/QtQml -I /usr/include/qt5/QtNetwork -I /usr/include/qt5/QtQuick -I /usr/include/qt5/QtQmlModels -I /usr/include/qt5/QtQuickControls2 -I /usr/include/qt5/QtSvg -I /usr/include/qt5/QtWidgets -DNDEBUG -fno-PIE -fno-PIE -mtune=i686 -O2    -ffile-prefix-map=/builddir/ktailctl-0.8.0/build=. -Wall -Wextra -pedantic -Wno-c99-extensions -Wno-c++98-compat -fno-operator-names -fno-exceptions -Wall -Wextra -Wcast-align -Wchar-subscripts -Wformat-security -Wno-long-long -Wpointer-arith -Wundef -Wnon-virtual-dtor -Woverloaded-virtual -Werror=return-type -Werror=init-self -Wvla -Wdate-time -Wsuggest-override -Wlogical-op -fdiagnostics-color=always -std=c++17 -flto=auto -fno-fat-lto-objects -fvisibility=hidden -fvisibility-inlines-hidden -fPIC -MD -MT src/CMakeFiles/ktailctl.dir/util.cpp.o -MF src/CMakeFiles/ktailctl.dir/util.cpp.o.d -o src/CMakeFiles/ktailctl.dir/util.cpp.o -c /builddir/ktailctl-0.8.0/src/util.cpp
../src/util.cpp: In function 'QString formatCapacityHumanReadable(long int)':
../src/util.cpp:42:56: error: narrowing conversion of '1099511627776' from 'long long int' to 'long int' [-Wnarrowing]
   42 |     static constexpr std::array<long, 5> limits = {1ll << 40ll, 1ll << 30ll, 1ll << 20ll, 1ll << 10ll, 1ll};
      |                                                    ~~~~^~~~~~~
```

**Comment by f-koehler** [comment]
@chrysos349 I forgot to change the type of the array in one of the functions. I added another fix to the aforementioned branch (also merged into `main` as of now). There is no need to disable the compiler warning.

If you have time, you can try to compile again. There should be no more issues...

**Comment by f-koehler** [comment]
Closed with #49

**Comment by chrysos349** [comment]
@f-koehler You're right! No more issues. Thanks for the fast fix!
